### PR TITLE
add config-option for an image's maximum filesize when generating previews

### LINKED ISSUE #13218: [6.0.6] Thumbnails of 140MB photo files result in a timeout error of the WebGUI
A user who uploaded a 140MB photo and couldn’t access it from the Web-interface to share or download it because the thumbnail creation failed with a timeout.

I couldn’t find a setting in config.php to limit this behavior like only create thumbnails of files smaller then X size.

we cannot increase the execution time in the php.ini because it’s already 30 seconds; increasing it will only slow down the web interface and annoy other users.

I looked in the documentation but there is no way to adjust the behavior of the thumbnail creation in ownCloud, so I need some kind of fix as such that we can influence the behavior.

For example, a max file size for thumbnails to be created.

@MorrisJobke 

00002256

### PR DESCRIPTION
**Pull request by georgehrke**
fixes https://github.com/owncloud/core/issues/13218

### CODE REVIEW

**Comment by MorrisJobke** [comment]
works :+1:

File: config/config.sample.php [hunk 1]
```diff
@@ -944,6 +944,17 @@
...
 /**
+ * max file size for generating image previews with imagegd (default behaviour)
+ * If the image is bigger, it'll try other preview generators,
+ * but will most likely show the default mimetype icon
+ *
+ * Value represents the maximum filesize in megabytes
+ * Default is 50
+ * Set to -1 for no limit
+ */
+'max_filesize_imagegd_previews' => 50,
```
Comment (DeepDiver1975) [inline config/config.sample.php]: typo

File: lib/private/preview/image.php [hunk 2]
```diff
@@ -26,6 +26,13 @@ public function getThumbnail($path, $maxX, $maxY, $scalingup, $fileview) {
 			return false;
 		}
 
+		$maxSizeForImages = \OC::$server->getConfig()->getSystemValue('max_filesize_imagegd_previews', 50);
```
Comment (DeepDiver1975) [inline lib/private/preview/image.php]: typo

Comment (georgehrke) [inline lib/private/preview/image.php]: it's supposed to say imagegd, or what typo do you mean?

Comment (DeepDiver1975) [inline lib/private/preview/image.php]: seriously - what is imagegd supposed to say?

Comment (georgehrke) [inline lib/private/preview/image.php]: php has image gd and imagick for handling images. Imagick doesn't cause issues. It's OC_Image which is based on image gd

Comment (DeepDiver1975) [inline lib/private/preview/image.php]: I'm totally aware of gd and imagick - but does the admin? Furthermore imagegd is a function we don't use.

from my pov we can simply get rid of the gd - or add another _

Comment (georgehrke) [inline lib/private/preview/image.php]: ok, we can also remove it, will take care

Comment (DeepDiver1975) [inline lib/private/preview/image.php]: > ok, we can also remove it, will take care

THX

File: config/config.sample.php [hunk 3]
```diff
@@ -944,6 +944,17 @@
 'forwarded_for_headers' => array('HTTP_X_FORWARDED', 'HTTP_FORWARDED_FOR'),
 
 /**
+ * max file size for generating image previews with imagegd (default behaviour)
```
Comment (DeepDiver1975) [inline config/config.sample.php]: @carlaschroder please review this documentation -THX

**Comment by carlaschroder** [comment]
:+1:

**Comment by MorrisJobke** [comment]
@oparoz @wakeup @sugaroverdose An easy to review PR ;) Simply use a big image ( https://commons.wikimedia.org/wiki/File:Bath_from_alexandra_park.jpg ) and test if it doesn't has a preview while smaller images still have previews. ;)

**Comment by oparoz** [comment]
My problem is that it doesn't really fix https://github.com/owncloud/core/issues/13218

I'd like this line to be fixed as part of the overall solution: https://github.com/owncloud/core/blob/master/lib/private/image.php#L470

And I'd like for these defaults to be modified to something reasonable like 2048x2048 so that we can start to generate sane previews for a browser/mobile environment.
https://github.com/owncloud/core/blob/master/config/config.sample.php#L611
https://github.com/owncloud/core/blob/master/config/config.sample.php#L616

File: config/config.sample.php [hunk 4]
```diff
@@ -944,6 +944,17 @@
...
+ * If the image is bigger, it'll try other preview generators,
+ * but will most likely show the default mimetype icon
+ *
+ * Value represents the maximum filesize in megabytes
+ * Default is 50
+ * Set to -1 for no limit
+ */
+'max_filesize_image_previews' => 50,
```
Comment (DeepDiver1975) [inline config/config.sample.php]: I just realized that we have other preview related settings - please move this one to the others and name accordingly

preview_max_filesize_image

**Comment by DeepDiver1975** [comment]
@oparoz we discussed already that fixing OC_Image is too critical for the current dev cycle - we can touch this class for OC8.1. This PR is a workaround for OC8

**Comment by oparoz** [comment]
@DeepDiver1975 Line 470 is a logging operation which kills PHP. Surely this can be removed in OC8 or simplified to just output a simple string?

The other settings I've mentioned have values of `null`, which make as much sense as having no limit on filesize imho.

**Comment by DeepDiver1975** [comment]
> @DeepDiver1975 Line 470 is a logging operation which kills PHP. Surely this can be removed in OC8 or simplified to just output a simple string?

Ah - I see what you mean now - sorry - I was on a different track .... :see_no_evil: 

It's indeed pointless to write the blob content to the log here : facepalm:

@georgehrke please submit a second pr where this log line is simply killed

@oparoz feel free to submit this pr as well if you are interested - THX

**Comment by georgehrke** [comment]
> @georgehrke please submit a second pr where this log line is simply killed

https://github.com/owncloud/core/pull/13614

**Comment by DeepDiver1975** [comment]
:+1:

**Comment by DeepDiver1975** [comment]
@karlitschek backport? Was reported as issue on oc7

**Comment by karlitschek** [comment]
backport would be good

**Comment by MorrisJobke** [comment]
stable7 2d2e024 - tested works

**Pull request merged**